src/ucp/tag/rndv.c

```diff
@@ -1189,49 +1540,54 @@ static ucs_status_t ucp_rndv_pipeline(ucp_request_t *sreq,
             freq->send.mdesc                      = NULL;
+
+            ucp_request_send(freq, 0);
         } else {
-            /* perform get on memtype endpoint to stage data to host memory */
-            mem_type_ep       = worker->mem_type_ep[sreq->send.mem_type];
-            mem_type_rma_lane = ucp_ep_config(mem_type_ep)->key.rma_bw_lanes[0];
-            if (mem_type_rma_lane == UCP_NULL_LANE) {
-                return UCS_ERR_UNSUPPORTED;
-            }
+            ucp_rndv_send_frag_get_mem_type(fsreq, 0, length, offset,
+                                            ucp_rndv_put_pipeline_frag_get_completion);
+        }
 
-            mdesc = ucp_worker_mpool_get(&worker->rndv_frag_mp);
-            if (mdesc == NULL) {
-                status = UCS_ERR_NO_MEMORY;
-                goto out;
-            }
+        offset += length;
+    }
 
-            ucp_request_send_state_init(freq, ucp_dt_make_contig(1), 0);
-            ucp_request_send_state_reset(freq, ucp_rndv_frag_get_completion,
-                                         UCP_REQUEST_SEND_PROTO_RNDV_GET);
-            md_index                              = ucp_ep_md_index(mem_type_ep, mem_type_rma_lane);
-            freq->send.ep                         = mem_type_ep;
-            freq->send.buffer                     = mdesc + 1;
-            freq->send.datatype                   = ucp_dt_make_contig(1);
-            freq->send.mem_type                   = sreq->send.mem_type;
-            freq->send.state.dt.dt.contig.memh[0] = ucp_memh2uct(mdesc->memh, md_index);
-            freq->send.state.dt.dt.contig.md_map  = UCS_BIT(md_index);
-            freq->send.length                     = length;
-            freq->send.uct.func                   = ucp_rndv_progress_rma_get_zcopy;
-            freq->send.rndv_get.rkey              = NULL;
-            freq->send.rndv_get.remote_address    =
-                    (uint64_t)UCS_PTR_BYTE_OFFSET(fsreq->send.buffer, offset);
-            freq->send.rndv_get.lanes_map         = 0;
-            freq->send.rndv_get.lane_count        = 0;
-            freq->send.rndv_get.rreq              = fsreq;
-            freq->send.mdesc                      = mdesc;
+    return UCS_OK;
+}
 
-        }
+UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_frag_rts_handler,
+                 (arg, data, length, flags),
+                 void *arg, void *data, size_t length, unsigned flags)
+{
+    ucp_worker_h worker                = arg;
+    ucp_rndv_frag_rts_hdr_t *rndv_frag_rts_hdr   = data;
+    ucp_request_t *rreq = (ucp_request_t *)rndv_frag_rts_hdr->rreq;
+    ucp_request_t *rndv_req;
 
-        ucp_request_send(freq, 0);
-        offset += length;
+    ucs_trace_req("recv_frag_rts rreq:0x%lx sreq:0x%lx length:%ld offset:%ld",
+                  rndv_frag_rts_hdr->rreq, rndv_frag_rts_hdr->sreq.reqptr,
+                  rndv_frag_rts_hdr->size, rndv_frag_rts_hdr->offset);
+
+    UCS_ASYNC_BLOCK(&worker->async);
```

Comment:
why need lock here?

Reply:
I don't think we need it.  I just got it from ucp_rndv_matched logic.